pkg/backup/backup/backup_cleaner.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backup
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap.com/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/backup/constants"
+	backuputil "github.com/pingcap/tidb-operator/pkg/backup/util"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	"github.com/pingcap/tidb-operator/pkg/label"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	batchlisters "k8s.io/client-go/listers/batch/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+)
+
+// BackupCleaner implements the logic for cleaning backup
+type BackupCleaner interface {
+	Clean(backup *v1alpha1.Backup) error
+}
+
+type backupCleaner struct {
+	statusUpdater controller.BackupConditionUpdaterInterface
+	secretLister  corelisters.SecretLister
+	jobLister     batchlisters.JobLister
+	jobControl    controller.JobControlInterface
+}
+
+// NewBackupCleaner returns a BackupCleaner
+func NewBackupCleaner(
+	statusUpdater controller.BackupConditionUpdaterInterface,
+	secretLister corelisters.SecretLister,
+	jobLister batchlisters.JobLister,
+	jobControl controller.JobControlInterface) BackupCleaner {
+	return &backupCleaner{
+		statusUpdater,
+		secretLister,
+		jobLister,
+		jobControl,
+	}
+}
+
+func (bc *backupCleaner) Clean(backup *v1alpha1.Backup) error {
+	if backup.DeletionTimestamp == nil {
+		// The backup object has not been deleted，do nothing
+		return nil
+	}
+	ns := backup.GetNamespace()
+	name := backup.GetName()
+
+	glog.Infof("start to clean backup %s/%s", ns, name)
+
+	cleanJobName := backup.GetCleanJobName()
+	_, err := bc.jobLister.Jobs(ns).Get(cleanJobName)
+	if err == nil {
+		// already have a clean job running，return directly
+		return nil
+	}
+
+	if backup.Status.BackupPath == "" {
+		// the backup path is empty, so there is no need to clean up backup data
+		return bc.statusUpdater.Update(backup, &v1alpha1.BackupCondition{
+			Type:   v1alpha1.BackupClean,
+			Status: corev1.ConditionTrue,
+		})
+	}
+	// not found clean job, create it
+	job, reason, err := bc.makeCleanJob(backup)
+	if err != nil {
+		bc.statusUpdater.Update(backup, &v1alpha1.BackupCondition{
+			Type:    v1alpha1.BackupFailed,
+			Status:  corev1.ConditionTrue,
+			Reason:  reason,
+			Message: err.Error(),
+		})
+		return err
+	}
+
+	if err := bc.jobControl.CreateJob(backup, job); err != nil {
```

Comment:
Do we need to delete the jobCleaner pod when finished?

Reply:
I think it's not necessary to delete the clean pod explicitly.  Because backup CR is the ower reference of the clean job. So when the backup CR is deleted, the clean job will also be GC by k8s automatically.